Project/Assets/ML-Agents/Examples/Worm/Scripts/WormAgent.cs

```diff
@@ -127,120 +163,81 @@ public override void CollectObservations(VectorSensor sensor)
-        bpDict[bodySegment3].SetJointTargetRotation(continuousActions[++i], continuousActions[++i], 0);
 
         // Update joint strength
+        bpDict[bodySegment0].SetJointStrength(continuousActions[++i]);
         bpDict[bodySegment1].SetJointStrength(continuousActions[++i]);
         bpDict[bodySegment2].SetJointStrength(continuousActions[++i]);
-        bpDict[bodySegment3].SetJointStrength(continuousActions[++i]);
 
-        if (bodySegment0.position.y < ground.position.y - 2)
+        //Reset if Worm fell through floor;
+        if (bodySegment0.position.y < m_StartingPos.y - 2)
         {
             EndEpisode();
         }
     }
 
     void FixedUpdate()
     {
-        if (detectTargets)
-        {
-            foreach (var bodyPart in m_JdController.bodyPartsDict.Values)
-            {
-                if (bodyPart.targetContact && bodyPart.targetContact.touchingTarget)
-                {
-                    TouchedTarget();
-                }
-            }
-        }
+        UpdateOrientationObjects();
 
-        // Set reward for this step according to mixture of the following elements.
-        if (rewardMovingTowardsTarget)
-        {
-            RewardFunctionMovingTowards();
-        }
+        var velReward =
+            GetMatchingVelocityReward(m_OrientationCube.transform.forward * m_maxWalkingSpeed,
+                m_JdController.bodyPartsDict[bodySegment0].rb.velocity);
 
-        if (rewardFacingTarget)
-        {
-            RewardFunctionFacingTarget();
-        }
+        //Angle of the rotation delta between cube and body.
+        //This will range from (0, 180)
+        var rotAngle = Quaternion.Angle(m_OrientationCube.transform.rotation,
+            m_JdController.bodyPartsDict[bodySegment0].rb.rotation);
 
-        if (rewardUseTimePenalty)
+        //The reward for facing the target
+        var facingRew = 0f;
+        //If we are within 30 degrees of facing the target
+        if (rotAngle < 30)
         {
-            RewardFunctionTimePenalty();
+            //Set normalized facingReward
+            //Facing the target perfectly yields a reward of 1
+            facingRew = 1 - (rotAngle / 180);
         }
-    }
 
-    /// <summary>
-    /// Reward moving towards target & Penalize moving away from target.
-    /// </summary>
-    void RewardFunctionMovingTowards()
-    {
-        m_MovingTowardsDot = Vector3.Dot(m_JdController.bodyPartsDict[bodySegment0].rb.velocity, m_DirToTarget.normalized);
-        AddReward(0.01f * m_MovingTowardsDot);
+        //Add the product of these two rewards
+        AddReward(velReward * facingRew);
     }
 
     /// <summary>
-    /// Reward facing target & Penalize facing away from target
+    /// Normalized value of the difference in actual speed vs goal walking speed.
     /// </summary>
-    void RewardFunctionFacingTarget()
+    public float GetMatchingVelocityReward(Vector3 velocityGoal, Vector3 actualVelocity)
```

Comment:
Okay, this is the third time I see this method. I think it should be static and moved to the shared assets folders so it is written only once. There are probably other methods that should follow.

Reply:
no change. I agree, I plan on doing a PR after the current ragdoll PRs are merged to consolidate some of the reused methods